From a4695eb4670fda06fc9f364543fdb578023d022b Mon Sep 17 00:00:00 2001
From: Codex <codex@openai.com>
Date: Tue, 29 Jul 2025 01:55:06 +0000
Subject: [PATCH] fix detectOS default return

---
 compiler/src/build.d | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/compiler/src/build.d b/compiler/src/build.d
index bc0d3d7..c2a77b6 100755
--- a/compiler/src/build.d
+++ b/compiler/src/build.d
@@ -1684,7 +1684,7 @@ Returns: a string from `{windows, osx,linux,freebsd,openbsd,netbsd,dragonflybsd,
 */
 string detectOS()
 {
-    /*version(Windows)
+    version(Windows)
         return "windows";
     else version(OSX)
         return "osx";
@@ -1701,11 +1701,11 @@ string detectOS()
     else version(Solaris)
         return "solaris";
     else version(PowerNex)
-        return "powernex";*/
-    version(AnonymOS)
+        return "powernex";
+    else version(AnonymOS)
         return "anonymos";
-    /*else
-        static assert(0, "Unrecognized or unsupported OS.");*/
+    else
+        static assert(0, "Unrecognized or unsupported OS.");
 }
 
 /**
-- 
2.43.0

